### PR TITLE
Add Registry partition_by option for duplicate registries

### PR DESCRIPTION
**Pull request by studzien**
## Summary

This PR adds a new `partition_by` option to Registry for duplicate registries, allowing users to choose between `:pid` (default) and `:key` partitioning strategies to optimize performance for different use cases.

## Background

This feature addresses performance concerns first noticed in phoenixframework/phoenix_pubsub#198, where Registry's PID-based partitioning wasn't optimal for workloads with many topics and relatively few subscribers per topic.

### New partition_by Option
- **`:pid` (default)**: Traditional behavior - partitions by process ID, suitable for few keys with many entries
- **`:key`**: New behavior - partitions by key, suitable for many keys with few entries per key

### Performance Optimization
- Key-based lookups with `partition_by: :key` now only need to check a single partition
- Reduces lookup complexity from O(partitions) to O(1) for key-based operations

## Changes

**lib/elixir/lib/registry.ex:**
- Added `partition_by` to `start_option` typespec
- Updated all internal functions to handle the new partition_by parameter
- Enhanced documentation with usage examples and performance guidance
- Updated `values/3`, `lookup/2`, and other functions to properly handle key-based partitioning

**lib/elixir/test/elixir/registry_test.exs:**
- Added comprehensive tests for both partitioning strategies
- Extended cleanup tests to cover all partitioning combinations
- Added specific tests for `partition_by: :key` and `partition_by: :pid`

### New Option
```elixir
Registry.start_link(
  keys: :duplicate,
  name: MyRegistry,
  partitions: System.schedulers_online(),
  partition_by: :key  # New option
)
```

### Usage Guidelines
- **Use `:pid` (default)**: When you have few keys with many entries (e.g., one topic with many subscribers)
- **Use `:key`**: When you have many keys with few entries each (e.g., many topics with few subscribers)

## Backward Compatibility

This change is fully backward compatible:
- Default behavior remains unchanged (`:pid` partitioning)
- Existing code continues to work without modification
- Only adds new functionality for those who opt into `:key` partitioning
- Only supported for `:duplicate` registries (validated at startup)

## Related

- phoenixframework/phoenix_pubsub#198 - Original performance issue identification

### CODE REVIEW

File: lib/elixir/lib/registry.ex [hunk 1]
```diff
@@ -255,7 +256,7 @@ defmodule Registry do
 
   defp whereis_name(registry, key) do
     case key_info!(registry) do
-      {:unique, partitions, key_ets} ->
+      {:unique, partitions, key_ets, _} ->
```
Comment (josevalim) [inline lib/elixir/lib/registry.ex]: Instead of adding a new element, I wonder if we should introduce a new internal type. So we have `:unique`, `:duplicate` and `:duplicate_by_key`.

Comment (josevalim) [reply lib/elixir/lib/registry.ex]: We could even have `:duplicate_by_pid` and `:duplicate_by_key`. Or, alternatively, have the type be `{:duplicate, :pid | :key}`. But basically there is no need to store more information that only applies to duplicates and not unique.

Comment (studzien) [reply lib/elixir/lib/registry.ex]: I think this makes sense, considering that the new key doesn't make sense for a `unique` registry. Just wondering if `:duplicate` should also have an explicit alias, something like `:duplicate_by_pid`

Comment (josevalim) [reply lib/elixir/lib/registry.ex]: Yeah, I think we commented at the same time. Let's go with either`:duplicate_by_pid` and `:duplicate_by_key` or `{:duplicate, partition_by}`. Your call. The second one will allow us to partially match when we don't care about the partition.

Comment (studzien) [reply lib/elixir/lib/registry.ex]: Sounds good. I'll also make sure that `:duplicate` keeps existing behaviour of `:duplicate_by_pid` (or `{:duplicate, :pid}`})

**Comment by studzien** [comment]
@josevalim I updated it to use `{:duplicate, :key}` or `{:duplicate, :pid`}. 
I was not certain if the partition supervisor strategy should change, but I think it should stay the same (`:one_for_one`) since we partition both ets tables by the same key (`:key` or `:pid`).

Will update the PR description shortly.

File: lib/elixir/lib/registry.ex [hunk 2]
```diff
@@ -467,7 +511,7 @@ defmodule Registry do
             :error
         end
 
-      {kind, _, _} ->
+      {kind, _, _, _} ->
```
Comment (josevalim) [inline lib/elixir/lib/registry.ex]: This claude needs to be reverted but we need to be careful with `#{kind}` below, that will no longer work.

File: lib/elixir/lib/registry.ex [hunk 3]
```diff
@@ -1559,7 +1621,9 @@ defmodule Registry.Supervisor do
       for i <- 0..(partitions - 1) do
         key_partition = Registry.Partition.key_name(registry, i)
         pid_partition = Registry.Partition.pid_name(registry, i)
-        arg = {kind, registry, i, partitions, key_partition, pid_partition, listeners, compressed}
+
+        arg =
+          {kind, registry, i, partitions, key_partition, pid_partition, listeners, compressed}
```
Comment (josevalim) [inline lib/elixir/lib/registry.ex]: Reverting:

```suggestion
        arg = {kind, registry, i, partitions, key_partition, pid_partition, listeners, compressed}
```

**Review by josevalim — COMMENTED** [review]
I have dropped a few comments.

I also think we need to better parameterize the test:

1. Break the current RegistryTest into Registry.UniqueTest and Registry.DuplicateTest
2. Make sure the DuplicateTest is parameterized by partition count AND partition by

The new tests you added can be part of the new DuplicateTest.

**Comment by studzien** [comment]
Excellent feedback, thank you!
I will have time to continue working on this around Wednesday/Thursday this week.

**Comment by josevalim** [comment]
Ping! Last call if you want this included in Elixir v1.19 :)

**Comment by studzien** [comment]
> Ping! Last call if you want this included in Elixir v1.19 :)

Thanks for the ping and sorry for the delay!
If I have this ready on Monday, can it still make the release cut?

**Comment by josevalim** [comment]
Yes!

**Comment by studzien** [comment]
1. I split the tests between unique and duplicate test modules. Running the fully parameterized tests on the duplicate test suite uncovered a bug in the `Registry.keys/2` function (for `{:duplicate, :keys}` we need to iterate over all partitions). I'm not happy with how the function now looks (it's a bit long), so please let me know if there's a pattern of shortening it that I could apply
2. I removed the tests I added earlier that covered successful registry startup for all permutations of kinds and partitions; parameterized tests now cover it
3. I reproduced the issue with the missed clause by adding a test for the `update_value` function in the duplicate registry.
4. I left the cleanup test code in the `RegistryTest`

**Comment by josevalim** [comment]
Beautiful work. It all looks great to me! It will be merged once tests pass.
